Various build / test / package imrovements: Use PyTest, build Python package, etc.

rjurney opened the pull request:
I did some improvements in the project's build and testing in the course of #473 that I have split out into this PR.

<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. Ensure you have added or run the appropriate tests for your PR
  2. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP]Your PR title ...'.
  3. Be sure to keep the PR description updated to reflect all changes.
  4. Please write your PR title to summarize what this PR proposes.
  5. If possible, provide a concise example to reproduce the issue for a faster review.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If there is design documentation, please add the link.
  3. If there is a discussion in the mailing list, please add the link.
-->



### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->

rjurney commented:
This has to go in before #511 or #513 will work... I refer to graphframes.tutorials which only makes sense if a package exists.

rjurney commented:
@SauronShepherd @SemyonSinchenko @bjornjorgensen plz take a look

SemyonSinchenko commented on python/setup.cfg:
Why not to use pyproject.toml instead? It seems to me that `setup.cfg` is a legacy approach.

rjurney replied:
Because setup.py is already there and I needed to wrap this quickly so I could see the `graphframes.tutorials` package from other Python scripts. A later PR can do that, I'm good at poetry.

rjurney replied:
@SemyonSinchenko okay, if I need extras I am going to poetrize it.

SemyonSinchenko replied:
@rjurney I did not mean moving to poetry! To be honest I was expecting that you leave setup.py as is and just move the package metadata to the pyproject file with backend "setuptools"... Are you 100% sure that we should use poetry here?

SemyonSinchenko replied:
https://packaging.python.org/en/latest/guides/writing-pyproject-toml/

rjurney replied:
Oh, yes I think we should move to poetry. It makes building and publishing a package really easy. It only affects developers who will be happier.

bjornjorgensen replied:
apache spark does not use poetry, but there are some mails about the project python build at @dev

rjurney replied:
Ah, thank you. I did not think about it. I do think we should try to use poetry, and deal with the merge when we merge... I don't want to rewrite this again. Poetry is 100% cleaner than the setup.py/cfg method I was just using. It was very easy, just got to fix CI. Is is also easy to publish to PyPi.

SemyonSinchenko replied:
My opinion is that if we are moving to a new tool we should move to `uv`. Or let's stay on `setuptools`.

SauronShepherd replied:
+1 for uv

SemyonSinchenko replied:
Ok, let's continue with poetry.

SemyonSinchenko commented on python/setup.cfg:
```diff
@@ -1,2 +1,42 @@
+include_package_data = True
+install_requires =
+    pyspark>=2.0.0
+    click==8.1.8
```
Why do we need it in main requirements?

rjurney replied:
Because `graphframes.tutorials.download` uses it. `Click` has no dependencies except for `colorama` on Windows, so it seems a safe inclusion. If you want me to pull it into an extras, I think that could confuse new users who try to run the tutorials...

SemyonSinchenko commented on python/setup.cfg:
```diff
@@ -1,2 +1,42 @@
+    click==8.1.8
+    numpy>=1.7
+    py7zr==0.22.0
+    requests==2.32.3
```
It seems to me that this should be a part of dev / additional dependencies

rjurney replied:
@SemyonSinchenko Okay, so I could do a `tutorials` extras. I can do that. It is more than one package it adds... alright :)

SemyonSinchenko commented on python/graphframes/tests.py:
```diff
@@ -15,63 +15,72 @@
+            "spark.submit.pyFiles",
+            os.path.abspath("python/dist/graphframes-{VERSION}-py3-none-any.whl"),
+        )
+        cls.sc = SparkContext(master="local[4]", appName="GraphFramesTests", conf=cls.conf)
```
Could we avoid it at all? The problem is that Spark Connect does not support SparkContext at all. And I don't see actual usage of `cls.sc`. Can we rely on the `SparkSession` only? As I remember it is also recommended by PySpark documentation too.

bjornjorgensen commented:
why do you add a dockerFile in the PR? 
The dockeFile have nothing to do with the rest of the PR. 

Convert tests to PyTest 

somewhere you will have things like Apache Spark does it 
like in this comment https://github.com/graphframes/graphframes/issues/504#issuecomment-2650100107
but now you will convert tests to pytests..

bjornjorgensen commented on python/graphframes/tests.py:
```diff
@@ -15,63 +15,72 @@
             raise TypeError("invalid minor version")
         try:
-            version_info['minor'] = int(m.group(2))
+            version_info["minor"] = int(m.group(2))
```
what is this?

bjornjorgensen commented on python/graphframes/tests.py:
```diff
@@ -123,28 +135,38 @@ def setUp(self):
+        assert gtu.spark_at_least_of_version("1.7")
+        assert gtu.spark_at_least_of_version("2.0")
+        assert gtu.spark_at_least_of_version("2.0.1")
+        assert gtu.spark_at_least_of_version("2.0.2")
```
what is this?

rjurney replied:
@bjornjorgensen I do not know, this is not an addition I made. It is not in my PR.

bjornjorgensen replied:
?

rjurney replied:
I converted it from unittest to pytest. I can alter the logic of the tests, but that was not the purpose of this PR.

bjornjorgensen commented on python/graphframes/tests.py:
```diff
@@ -376,93 +413,101 @@ def test_parallel_personalized_page_rank(self):
+        # Create bidirectional edges.
         all_edges = [z for (a, b) in edges for z in [(a, b), (b, a)]]
-        edges = self.spark.createDataFrame(all_edges, ["src", "dst"])
+        edgesDF = self.spark.createDataFrame(all_edges, ["src", "dst"])
```
this is not the same 
edges != edgesDF

rjurney replied:
Sorry, you don't like the use of `edgesDF` variable name? The edges are defined right there a few lines above.

rjurney replied:
@bjornjorgensen fixed

rjurney commented:
> why do you add a dockerFile in the PR? The dockeFile have nothing to do with the rest of the PR.
> 
> Convert tests to PyTest

@bjornjorgensen I don't add a Dockerfile. It was already there, I just cleaned it up a little.

> somewhere you will have things like Apache Spark does it like in this comment [#504 (comment)](https://github.com/graphframes/graphframes/issues/504#issuecomment-2650100107) but now you will convert tests to pytests..

I didn't want to include formatting in this PR as I need to get the other stuff in before hackathon.

rjurney commented:
Okay, if I have to do a `tutorials` extras I am just going to poetrize the whole thing...

bjornjorgensen commented:
@rjurney have 1 (one) PR for each thing.. 

not 1 for 40 half finished stuff..

rjurney commented:
> @rjurney have 1 (one) PR for each thing..
> 
> not 1 for 40 half finished stuff..

I'm going to finish them but obviously need help in reviewing the code. This wasn't easy to pull apart as the tutorials depends on the existence of a Python package.

bjornjorgensen commented:
> > @rjurney have 1 (one) PR for each thing..
> > not 1 for 40 half finished stuff..
> 
> I'm going to finish them but obviously need help in reviewing the code. This wasn't easy to pull apart as the tutorials depends on the existence of a Python package.

This takes a lot of time when you are posting PR's with over 1000 lines of code changes and.. no.. 
as i say in yours last PR start with one thing like updating the dockerFile and nothing more in that PR.. 
We can then revue it easier and it needed we can revers it to..

rjurney commented:
> > > @rjurney have 1 (one) PR for each thing..
> > > not 1 for 40 half finished stuff..
> > 
> > 
> > I'm going to finish them but obviously need help in reviewing the code. This wasn't easy to pull apart as the tutorials depends on the existence of a Python package.
> 
> This takes a lot of time when you are posting PR's with over 1000 lines of code changes and.. no.. as i say in yours last PR start with one thing like updating the dockerFile and nothing more in that PR.. We can then revue it easier and it needed we can revers it to..

I'm a little confused by your comments - this PR is small other than fairly simple changes to convert unittest to pytest, which had to be done in one go. You're reviewing code I did not write.

rjurney commented:
@bjornjorgensen okay, I see your point there are a few things going on here. I would ask for your patience - I tried to do one simple thing, create a graphframes motif finding tutorial and update the README to contain helpful content... it was bare. I added references to the motif finding tutorial and cleaned up the docs generally. This was intended to be one PR.

Then the following happened:

1. I could not import GraphFrames from my tutorial, because there was no actual package. This has always been a weird thing about GraphFrames. So I setup a package. I'm redoing this in poetry now, as a separate PR.
3. I couldn't use GraphFrames with a modern Python because `nose` required Python 3.9. It is very old and a very bad package. So I removed it and converted the unit tests to pytest. This was done in one go.
4. My branch would not build for unknown reasons. It wasn't clear from the error messages, so I fiddled with versions and things until it built. That is how you wind up with the stuff in this PR.

bjornjorgensen commented:
In this PR you have 
- upgraded version 

- div upgrades in CI like upgraded python, scala. 
- something with requierments-dev 

- added files and folder to gitignore 

- Updated dockerFile

- Converted nose testes to pytests 

- build a python package 

- removed python2 from run tests script 

- add a new function parse_requirements 

- changed python sys to os for file operations 

Have one PR for each of them.

rjurney commented:
@bjornjorgensen can you please refresh and look at my edited / updated comments? I agree that would be the way to go, but it does build and the hackathon looms.

rjurney commented:
> In this PR you have
> 
> * upgraded version
> * div upgrades in CI like upgraded python, scala.
> * something with requierments-dev
> * added files and folder to gitignore
> * Updated dockerFile
> * Converted nose testes to pytests
> * build a python package
> * removed python2 from run tests script
> * add a new function parse_requirements
> * changed python sys to os for file operations
> 
> Have one PR for each of them.

A lot of this stuff is all part of building a package. Don't you see the unifying theme of these things?

> * something with requierments-dev

This is so we don't install the dev dependencies when we build the package.

> * added files and folder to gitignore

These are files created by the package build process.

> * Updated dockerFile

Removed.

> * Converted nose testes to pytests

Okay, I hear you. I will pull this out.

Okay, I am going to do what you ask... it's just very time consuming.

bjornjorgensen commented:
This is a 10 years old project and we need to upgrade it. 
Working with git means you should be able to revert PR's easy to. 

If you are posting PR's with only one thing changed then it will be much easier for others to spend time and review changes. And in the new code we need to know why things are there to.

rjurney commented on .github/workflows/python-ci.yml:
```diff
@@ -7,8 +7,8 @@ jobs:
           - spark-version: 3.5.4
-            scala-version: 2.12.18
-            python-version: 3.9.19
+            scala-version: 2.12.20
```
@bjornjorgensen I backed a lot of stuff out. Now this is the only unrelated commit to building a `graphframes` package.

rjurney commented:
@bjornjorgensen I had missed some stuff. NOW everything is required in one go to build a package.

rjurney commented:
@bjornjorgensen okay, check out the tests. There is the state of things where I am completely stuck. I don't know how to fix this build error without making various changes. When I upgraded the Python, shit "just worked."

rjurney commented:
@bjornjorgensen sorry, give me a minute. I included some dev stuff we don't need - pytest.

bjornjorgensen commented on python/requirements.txt:
```diff
@@ -1,3 +1,4 @@
 # This file should list any python package dependencies.
 nose==1.3.7
+pyspark>=2.0.0
```
graphframes have droped support for spark below version 3.4 or something

rjurney replied:
3.4 is pretty recent - shouldn't we keep it as broad as we can?

bjornjorgensen replied:
so users with spark 2.4 should try it to figure out that we don't support it?

rjurney replied:
When you say we don't support <3.4, where did that happen? I thought we supported more of a range than that.

bjornjorgensen replied:
https://github.com/graphframes/graphframes/pull/430

rjurney replied:
Okay, sorry I forgot. Updated the file.

bjornjorgensen commented on python/setup.cfg:
```diff
@@ -1,2 +1,42 @@
-# This file contains the default option values to be used during setup. An
-# example can be found at https://github.com/pypa/sampleproject/blob/master/setup.cfg
+[metadata]
```
should this file be named pyproject.toml ?

rjurney replied:
I am converting it to use a `pyproject.toml` right now.

Note that I reserved us the space: https://pypi.org/project/graphframes-py/

rjurney commented:
@bjornjorgensen any clue why the scala CI tests stopped working? I made no changes to any of that stuff... this is confusing me.

rjurney commented:
Not sure but am testing the build with 2.13.8 like in CI...

rjurney commented:
@bjornjorgensen this is all logically joined at once to create a `graphframes-py` package. It would break things to do it one file at a time. I hope it isn't overwhelming, I tried to keep it minimal. The one thing I did that is optional is the settings to make the dev dependencies work. I'd like to get the dev deps in this PR.

rjurney commented:
Okay... Scala failure was transient :D Now to fix Python!

rjurney commented:
@bjornjorgensen oh, re: poetry... I need to create extras like `graphframes[tutorials]` due to feedback on that PR. The old `setup.py` with the old build tools can't support extras without `pip`. I would rather adopt `poetry` than `pip` if we have to choose one.

rjurney commented:
@bjornjorgensen this is getting closer to ready for review... let me know what you think :) I could use @SauronShepherd and @SemyonSinchenko take as well.

SemyonSinchenko commented on .github/workflows/python-ci.yml:
```diff
@@ -31,12 +31,14 @@ jobs:
-        python -m pip install --upgrade pip wheel
-        pip install -r ./python/requirements.txt
-        pip install pyspark==${{ matrix.spark-version }}
+        python -m pip install --upgrade poetry
```
I would recommend to use [action](https://github.com/snok/install-poetry/) instead.

SemyonSinchenko commented on .github/workflows/python-ci.yml:
```diff
@@ -31,12 +31,14 @@ jobs:
       run: |
-        export SPARK_HOME=$(python -c "import os; from importlib.util import find_spec; print(os.path.join(os.path.dirname(find_spec('pyspark').origin)))")
-        ./python/run-tests.sh
+        export SPARK_HOME=$(poetry run python -c "import os; from importlib.util import find_spec; spec = find_spec('pyspark'); print(os.path.join(os.path.dirname(spec.origin)))")
```
Why do we need it? Tests will work even without SPARK_HOME

SemyonSinchenko commented on python/pyproject.toml:
```diff
@@ -0,0 +1,48 @@
+[tool.poetry]
+name = "graphframes-py"
+version = "0.8.4"
```
Let's use something like this: https://pypi.org/project/poetry-dynamic-versioning/?

SemyonSinchenko commented on python/pyproject.toml:
```diff
@@ -0,0 +1,48 @@
+[tool.poetry.group.dev.dependencies]
+black = "^25.1.0"
+flake8 = "^7.1.1"
+isort = "^6.0.0"
```
It looks like `pytest` is missing.

SemyonSinchenko reviewed (approved):
LGTM! I think there is still some place for improvements, but we can do it in next PRs. Thank you @rjurney !

rjurney commented:
Thanks!
